doc/source/reference/c-api.types-and-structures.rst

```diff
@@ -696,15 +698,29 @@ PyUFunc_Type
           PyUFuncGenericFunction *functions;
           void **data;
           int ntypes;
-          int reserved1;
+          int version;
           const char *name;
           char *types;
           const char *doc;
           void *ptr;
           PyObject *obj;
           PyObject *userloops;
+          int core_enabled;
+          int core_num_dim_ix;
+          int *core_num_dims;
+          int *core_dim_ixs;
+          int *core_offsets;
+          char *core_signature;
+          PyUFunc_TypeResolutionFunc *type_resolver;
+          PyUFunc_LegacyInnerLoopSelectionFunc *legacy_inner_loop_selector;
+          void *reserved2;
```

Comment:
This member right here is our get-out clause to avoid ABI issues. Instead of making the struct any longer, we can use this to point to an arbitrary and private `ufunc_internal` object allocated on the heap. With that, we become free to add new members at any point

Reply:
I'm not sure it is necessarily best to keep things private; I think ufunc writers would like to have access to whether dimensions are frozen, etc., in case the write, e.g., their own type resolver. (I've certainly made use of other numbers inferred from the signature in ufunc code.)
But as a variant on this: we could let this point to a new explicitly defined object that is version dependent (and gets a version-dependent name).

Reply:
> I think ufunc writers would like to have access to [...]

Can we use the same approach as we do for ndarray, where we provide accessor functions?

Reply:
adopt a form of this idea, which achieves the goals of allowing memcpy of the old size while still using a single malloc

Reply:
Hmm, thinking again I think there's no safe way to do this if someone uses a `memcpy`...